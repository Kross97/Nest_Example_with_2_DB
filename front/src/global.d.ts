
// обязательная установка sass пакета (npm i sass --save-dev)

declare module '*.module.scss' {
  const css: Record<string, string>;
  export default css;
}

declare module '*.scss' {
  const css: Record<string, string>;
  export default css;
}