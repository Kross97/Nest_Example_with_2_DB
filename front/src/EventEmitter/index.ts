interface IHandler {
  once: boolean;
  handler: () => void;
}

export const EventsDictionary = {
  unAuthorized: "unAuthorized",
} as const;

type TEvents = keyof typeof EventsDictionary;

class EventEmitter {
  private eventsHandlersDispatcher: Record<string, IHandler[]> = {};

  add(event: TEvents, handler: () => void, once?: boolean) {
    if (event in this.eventsHandlersDispatcher) {
      this.eventsHandlersDispatcher[event].push({ handler, once: !!once });
    } else {
      this.eventsHandlersDispatcher[event] = [{ handler, once: !!once }];
    }
  }

  remove(event: TEvents, handler: () => void) {
    if (event in this.eventsHandlersDispatcher) {
      this.eventsHandlersDispatcher[event] = this.eventsHandlersDispatcher[event].filter(
        (handlerObj) => handlerObj.handler !== handler
      );
    }
  }

  emit(event: TEvents) {
    if (event in this.eventsHandlersDispatcher) {
      this.eventsHandlersDispatcher[event] = this.eventsHandlersDispatcher[event].filter(
        (handlerObj) => {
          handlerObj.handler();
          return !handlerObj.once;
        }
      );
    }
  }
}

export const EventEmitterAgent = new EventEmitter();
