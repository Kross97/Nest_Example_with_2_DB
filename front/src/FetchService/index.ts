import {  getTokenJwt } from '../provider/AllProvider';
import { errorHandler } from './utils/errorHandler';

interface IUrlParams {
  url: string;
  headers?: Record<string, string>
  body?: any;
  requestType?: 'none' | 'json';
  responseType?: 'json' | 'blob' | 'text' | 'arrayBuffer';
  method: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';
}

const dispatcherHeaders = {
  'json': { 'content-type': 'application/json; charset=utf-8' },
  'none': {}
};

const dispatcherBody = {
  'json': (body: any) => JSON.stringify(body),
  'none': (body: any) => body,
}

const initAuthHeader = () => ({ 'Authorization': `Bearer ${getTokenJwt()}`})

class FetchService {

  private backUrl: string = process.env.BACK_HOST || 'http://localhost:3001';

  async request<T = any>({ url, headers = {}, body = '', method, requestType = 'none', responseType = 'json' }: IUrlParams) {
    const urlHost = (this.backUrl + (url.startsWith('/') ? url : `/${url}`));
    const currentHeaders = {...headers, ...dispatcherHeaders[requestType], ...initAuthHeader()};
    const currentUrl = method === 'GET' ? `${urlHost}?${new URLSearchParams(body).toString()}` : urlHost;

    const response = await fetch(currentUrl, {
      headers: currentHeaders,
      method,
      body: method === 'GET' ? null : dispatcherBody[requestType](body),
      credentials: 'include'
    });

    if(response.ok) {
      const contentLength = Number(response.headers.get('content-length') || 0);
      if (contentLength) {
        return response[responseType]();
      }
    } else {
      errorHandler(response);
    }
  }

  async postRequest<T = any>(data: Omit<IUrlParams, 'method'>) {
    return this.request<T>({ ...data, method: 'POST' });
  }

  async getRequest<T = any>(data: Omit<IUrlParams, 'method'>) {
    return this.request<T>({ ...data, method: 'GET' });
  }

  async putRequest<T = any>(data: Omit<IUrlParams, 'method'>) {
    return this.request<T>({ ...data, method: 'PUT' });
  }
};

export const FetchAgent = new FetchService();