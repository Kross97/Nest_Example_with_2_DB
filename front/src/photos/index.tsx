import cn from './Photos.module.scss';
import React, { SyntheticEvent, useEffect, useState } from 'react';
import { FetchAgent } from '../FetchService';

interface IFile { id: number; originalname: string; mimetype: string;}

export const PhotoBlock = () => {
  const [file, setFile] = useState<File | null | undefined>(null);
  const [files, setFiles] = useState<IFile[]>([]);

  useEffect(() => {
    FetchAgent.getRequest({ url: '/photos/all' }).then((results) => {

      setFiles(results);
    });
  }, []);


  const change = (e: SyntheticEvent) => {
    const file = (e.target as HTMLInputElement).files?.[0];
    const formData = new FormData();
    if (file) {
      formData.set('file', file);
      FetchAgent.postRequest({ url: '/photos', body: formData });

    }
  };

  const changeUserPhoto = (e: SyntheticEvent) => {
    const file = (e.target as HTMLInputElement).files?.[0];
    setFile(file);
  };
  const createUser = () => {
    if (file) {
      const formData = new FormData();
      formData.append('file', file);
      formData.set('user', JSON.stringify({
        name: {
          first: 'front_photo_first_name',
          last: 'front_photo_last_name',
        }, test: 'test ', car: {
          model: 1,
          year: 2001,
        },
      }));

      FetchAgent.postRequest({ url: '/user', body: formData });

    }
  };
  const downloadFile = async (file: IFile) => {
    const arrayBuffer = await FetchAgent.getRequest({ url: `/photos/download/first/${file.id}`, responseType: 'arrayBuffer'});
    const blob = new Blob([arrayBuffer], {
      type: file.mimetype,
    });

    const a = document.createElement('a');
    a.href = URL.createObjectURL(blob);
    a.download = file.originalname;
    a.click();
  };

  return <div className={cn.photosBlock}>
    <h3>Фотографии</h3>
    <div className={cn.column}>
      <span>Закачка фотографии</span>
      <input onChange={change} type={'file'} />
    </div>
    <div className={cn.column} style={{ alignItems: 'flex-start'}}>
      <span>Установка пользователя вместе с фотографией</span>
      <input onChange={changeUserPhoto} type={'file'} />
      { file && <button onClick={createUser}>Создать пользователя</button>}
    </div>
    <div className={cn.listFiles}>
      {files.map((f: any) => {
        return <div>
          <span>{f.originalname}</span>
          <button onClick={() => downloadFile(f)}>загрузка</button>
        </div>;
      })}
    </div>
  </div>;
};