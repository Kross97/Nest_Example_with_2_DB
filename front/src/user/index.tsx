import cn from './User.module.scss';
import React, { ChangeEvent, useEffect, useState } from 'react';
import { FetchAgent } from '../FetchService';
import { IUser } from './types';
import { ModalUser } from './ModalUser';


export const UserBlock = () => {
  const [userEdit, setUserEdit] = useState<IUser | null>(null)
  const [users, setUsers] = useState<IUser[]>([]);
  const [data, setUser] = useState({});

  const showEditHandler = (user: IUser | null) => setUserEdit(user);

  useEffect(() => {
    FetchAgent.getRequest({ url: '/user/all'}).then((results) => {
      setUsers(results);
    })
  }, []);

  const change = (e: ChangeEvent<HTMLInputElement>) => {
   const nameField = e.target.name;
   const value = e.target.value;
   setUser((prev) => ({...prev, [nameField]: value }));
  };

  const submitUser = async () => {
    await FetchAgent.postRequest({ url: '/user/create', body: data })
    setUser({});
  }

  const refreshUsers = async () => {
    await FetchAgent.getRequest({ url: '/user/refresh'})
  };

  return <>
  <ModalUser user={userEdit} showEditHandler={showEditHandler}/>
    <div className={cn.userBlock}>
      <div className={cn.formUser}>
        <span>Name</span>
        <input onChange={change} type={'text'} name={'name'} />
        <span>Логин</span>
        <input onChange={change} type={'text'} name={'login'} />
        <span>Пароль</span>
        <input onChange={change} type={'password'} name={'password'}/>
        <button onClick={submitUser}>Создать пользователя</button>
      </div>
      <button onClick={refreshUsers}>рефреш_логинов</button>
      <div style={{ display: 'flex', rowGap: '12px', flexDirection: 'column'}}>
        {users.map((userList) => <div>
          <span>{`${userList.name.first} ${userList.name.last}`}</span>
          <button onClick={() => showEditHandler(userList)}>изменить</button>
        </div>)}
      </div>
    </div>
  </>
};