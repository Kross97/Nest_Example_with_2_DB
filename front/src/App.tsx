import React from 'react';
import './App.css';
import { PhotoBlock } from './modules/photos';
import { UserBlock } from './modules/user';
import { AuthBlock } from './modules/auth';
import {  useAllContext } from './provider/AllProvider';
import { CryptoBlock } from './modules/crypto';

function App() {
  const { token, setToken } = useAllContext();

  const exitHandler = () => {
    setToken(null);
  };

  return (
    <>
      {token && <button onClick={exitHandler} style={{ position: 'fixed', top: '50px', right: '50px'}}>exit</button>}
      <div style={{ display: 'flex'}}>
        {token && <PhotoBlock />}
        {token && <UserBlock />}
        {token && <CryptoBlock />}
        {!token && <AuthBlock />}
      </div>
    </>
  );
}

export default App;
