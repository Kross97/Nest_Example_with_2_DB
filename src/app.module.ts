import { Module } from '@nestjs/common';
import { AppController } from './app.controller';
import { AppService } from './app.service';
import { TypeOrmModule } from '@nestjs/typeorm';
import {ConfigModule, ConfigService } from '@nestjs/config';
import { UserModule } from './modules/user/user.module';
import { RentCarsModule } from './modules/rentCars/rentCars.module';
// import dbConfiguration from "./config/db.config";


// способ 1
// https://stackoverflow.com/questions/59913475/configure-typeorm-with-one-configuration-for-cli-and-nestjs-application

@Module({
  imports: [TypeOrmModule.forRoot({
    type: "postgres",
    host: 'localhost',
    port: 5430,
    username: 'postgres_kross97',
    password: 'postgres_kross97',
    database: 'postgres_db_keross97',
    synchronize: false,
    migrationsRun: false,
    autoLoadEntities: false,
    migrationsTableName: "kross97_migration_table",
    entities: ['dist/entities/**/*.entity.js'],
    migrations: ['dist/migrations/*.js'],
  }), ConfigModule.forRoot(), UserModule, RentCarsModule],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {}
