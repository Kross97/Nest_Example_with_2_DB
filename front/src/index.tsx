import React from 'react';
import ReactDOM from 'react-dom/client';
import './index.css';
import { FullApp } from './App';
import { AllProvider } from './provider/AllProvider';
import './FetchService/index';

const root = ReactDOM.createRoot(
  document.getElementById('root') as HTMLElement
);
root.render(
  <React.StrictMode>
    <AllProvider>
      <FullApp />
    </AllProvider>
  </React.StrictMode>
);
