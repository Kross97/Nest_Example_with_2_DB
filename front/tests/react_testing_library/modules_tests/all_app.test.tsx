import React from "react";
import { fireEvent, render, screen, act } from "test-utils";
import { FullApp } from "../../../src/App";
import { FetchAgent } from "../../../src/FetchService";
import "@testing-library/jest-dom";
import { waitFor, waitForElementToBeRemoved } from "@testing-library/dom";

const headerClusterBlock = "Работа с кластеризацией сервера (Cluster модуль node)";

beforeAll(() => {
  // console.log('all', FetchAgent.postRequest);
});

test("Тестирование авторизации, смена страниц", async () => {
  const { container, baseElement } = render(<FullApp />);

  expect(screen.getByTestId("auth_header")).toBeInTheDocument();
  expect(screen.queryByText(headerClusterBlock)).not.toBeInTheDocument();

  const mockPost = jest.spyOn(FetchAgent, "postRequest").mockImplementation(() =>
    Promise.resolve({
      token: "ttt",
      currentPort: 3001,
    })
  );

  fireEvent.change(screen.getByLabelText("Логин"), { target: { value: "Kross97" } });

  const inputPassword = container.querySelector('[name="password"]');
  if (inputPassword) {
    fireEvent.change(inputPassword, { target: { value: "123" } });
  }

  /**
   * waitForElementToBeRemoved нужно получить реальный элемент (который счас отрисован)
   * чтобы он мог корректно за ним следить, когда он пропадет
   * иначе выпадает ошибка промиса
   * */
  // waitForElementToBeRemoved первый  аргумент должен быть элементом, массивом элементов или функцией обратного вызова,
  // которая возвращает элемент или массив элементов.
  waitForElementToBeRemoved(() => screen.queryByTestId("auth_header"), {
    timeout: 2000,
    interval: 50,
  }).then(() => {
    expect(screen.queryByText(headerClusterBlock)).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  /**
   * ACT FACT 2
   *
   * в act нужно оборачивать действия которые в коде могу повлечь за собой перендер компонентов
   *
   * render, fireEvent  - также включат внутри себя act под каппотом
   * */
  await act(async () => {
    fireEvent.click(screen.getByText("Войти"));
  });

  expect(mockPost).toHaveBeenCalled();
  await expect(mockPost.mock.results[0].value).resolves.toEqual({
    token: "ttt",
    currentPort: 3001,
  });

  const i = 1;
  await waitFor(
    () => {
      const authHeader = screen.queryByTestId("auth_header");
      // console.log(`попытки waitFor ${i++}`);
      if (!authHeader) {
        expect(authHeader).not.toBeInTheDocument();
      } else {
        /**
         * ошибка нужна чтобы коллбэк вызвался еще раз
         * */
        throw "для потворного колбэка waitFor";
      }
    },
    {
      timeout: 2000,
      interval: 50,
    }
  );
});
