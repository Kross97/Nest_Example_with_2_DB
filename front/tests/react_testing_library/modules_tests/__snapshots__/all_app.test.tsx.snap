// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тестирование авторизации, смена страниц 1`] = `
<body>
  <div>
    <button
      style="position: fixed; top: 50px; right: 50px;"
    >
      exit
    </button>
    <div
      style="display: flex; flex-wrap: wrap;"
    >
      <div
        class="clusterBlock"
      >
        <h2>
          Работа с кластеризацией сервера (Cluster модуль node)
        </h2>
        <h3>
          Состояние кластеризации: Включена
        </h3>
        <h3>
          Текущий порт: 3001
        </h3>
        <div>
          <span>
            Доступные порты:
          </span>
          <span />
        </div>
        <div>
          <button>
            Выбрать новый порт сервера
          </button>
          <button
            disabled=""
          >
            Включить кластеризацию
          </button>
          <button>
            Отключить кластеризацию
          </button>
        </div>
      </div>
      <div
        style="display: flex;"
      >
        <div
          class="photosBlock"
        >
          <h3>
            Модуль для работы с медиа-материалами (Сохранение и получение медиа из БД)
          </h3>
          <h3>
            Фотографии
          </h3>
          <div
            class="column"
          >
            <span>
              Закачка медиа
            </span>
            <input
              type="file"
            />
          </div>
          <div
            class="column"
          >
            <span>
              Закачка медиа (множественное)
            </span>
            <input
              multiple=""
              type="file"
            />
          </div>
          <div
            class="column"
            style="align-items: flex-start;"
          >
            <span>
              Установка пользователя вместе с фотографией
            </span>
            <input
              type="file"
            />
          </div>
          <label
            style="display: flex; flex-direction: column; row-gap: 10px;"
          >
            <b>
              Поиск
            </b>
            <input
              type="text"
            />
          </label>
          <div
            class="listFiles"
          />
        </div>
        <div
          class="userBlock"
        >
          <button
            style="margin-bottom: 10px;"
          >
            рефреш_логинов
          </button>
          <div
            style="padding: 10px; border: 2px solid black; margin-bottom: 14px;"
          >
            <h3>
              Модуль для работы с пользователями (Сохранение и получение из БД)
            </h3>
            <h4>
              Работа с отношениями 
              <br />
              1.один-к-одному, 
              <br />
              2.многие-ко-многим, 
              <br />
              3.один-ко-многим
            </h4>
            <div
              class="formUser"
            >
              <span>
                Name
              </span>
              <input
                name="nameFirst"
                type="text"
                value=""
              />
              <span>
                LastName
              </span>
              <input
                name="nameLast"
                type="text"
                value=""
              />
              <span>
                Логин
              </span>
              <input
                name="login"
                type="text"
                value=""
              />
              <span>
                Пароль
              </span>
              <input
                name="password"
                type="password"
                value=""
              />
              <div
                style="border: 1px solid black; padding: 10px; display: flex; flex-direction: column; align-items: center; row-gap: 16px; text-align: center; margin-bottom: 14px;"
              >
                <b>
                  Данный блок задействует создание 
                  <u>
                    user
                  </u>
                   полностью через FormData 
                  <u>
                    (и его кастомный парсинг на бэке)
                  </u>
                </b>
                <b>
                  вместе с множеством прикрепляемых фотографтий
                </b>
                <label>
                  Фото для пользователя (через formData)
                  <input
                    multiple=""
                    type="file"
                  />
                </label>
                <button>
                  Создать пользователя_(через_formData)
                </button>
              </div>
              <button>
                Создать пользователя_(через_json)
              </button>
            </div>
          </div>
          <label>
            Поиск
            <input
              type="text"
            />
          </label>
          <div
            class="listUsers"
          />
        </div>
      </div>
      <div
        style="display: flex; column-gap: 10px; white-space: wrap; word-break: break-all;"
      >
        <div
          class="cryptoBlock"
        >
          <h3>
            Модуль для работы с шифрованием (Crypto node.js)
          </h3>
          <h4>
            Шифрование и де-шифрование данных
          </h4>
          <div
            class="blockInfo"
          >
            <span>
              <b>
                1. "+$+"
              </b>
               - в шифрованном тексте является разделителем между вектором инициализации 
              <b>
                (iv)
              </b>
               и шифруемыми данными
            </span>
            <span>
              <b>
                2. iv (вектор инициализации)
              </b>
               - участвует в шифровании данных, при каждом новом шифровании 
              <b>
                (Cipher)
              </b>
               возвращаемый шифр будет отличаться т.к создается новый экземпляр iv (вектора инициализации)
            </span>
            <span>
              <b>
                3. iv берется из шифра по разделителю
              </b>
               - при каждой де-шифрации, что гарантирует корректную работу де-шифрации 
              <b>
                (Decipher)
              </b>
            </span>
          </div>
          <label
            class="column"
          >
            Данные для шифрования 
            <b>
              (класс Cipher node)
            </b>
            <input
              type="text"
              value=""
            />
          </label>
          <div
            class="column"
          >
            <span>
              Данные после шифрования 
              <b>
                (класс Decipher node)
              </b>
            </span>
            <span
              data-testid="cipher-field"
            />
          </div>
          <label
            class="column"
          >
            Данные для де-шифрования (hex - строка) 
            <b>
              (класс Decipher node)
            </b>
            <input
              type="text"
              value=""
            />
          </label>
          <div
            class="column"
          >
            <span>
              Данные после де-шифрования
            </span>
            <span
              data-testid="decipher-field"
            />
          </div>
          <button
            data-testid="cipher-button"
          >
            шифровать
          </button>
          <button
            data-testid="decipher-button"
          >
            де-шифровать
          </button>
        </div>
        <div
          class="cryptoSignBlock"
        >
          <h3>
            Модуль для работы с подписями (Sign, Verify node.js)
          </h3>
          <h4>
            Подпись и проверка данных
          </h4>
          <label
            class="column"
          >
            Данные для подписи 
            <b>
              (класс Sign, Verify node)
            </b>
            <input
              type="text"
              value=""
            />
          </label>
          <button>
            тест подписи
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
